### PR TITLE
TE-3614 Sort the queries for P&S

### PR DESCRIPTION
**Pull request by m7moud**
Includes #29

- Developer(s): @m7moud

- Ticket: https://spryker.atlassian.net/browse/TE-3614

- Academy PR: -

- Release Group: https://release.spryker.com/release-groups/view/1445

#### Please confirm

- [X] No new OS components - or they have been approved by legal department

#### Documentation

- [X] Functional documentation provided or in progress?
- [X] Integration guide for projects provided or not needed?
- [X] Migration guides for all contained majors provided or not needed?

#### Release Table

   Module                | Release Type         | Constraint Updates         |
   :--------------------- | :------------------------ | :--------------------- |
   EventBehavior         | minor                 |                       |

#### Release Notes

{skip}

-----------------------------------------

##### Change log

Improvements

- Introduced `EventResourceQueryContainerPluginIterator` to be responsible for iterating over the `EventResourceQueryContainerPluginInterface` plugins.
- Introduced `EventResourceRepositoryPluginIterator` to be responsible for iterating over the `EventResourceRepositoryPluginInterface` plugins.
- Introduced `EventResourceRepositoryBulkPluginIterator` to be responsible for iterating over the `EventResourceBulkRepositoryPluginInterface` plugins.
- Added `orderBy` to the query to ensure proper chunking.

### CODE REVIEW

File: src/Spryker/Zed/EventBehavior/Business/Model/EventResourceQueryContainerPluginIterator.php [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+use Iterator;
+use Propel\Runtime\ActiveQuery\ModelCriteria;
+use Spryker\Zed\EventBehavior\Dependency\Plugin\EventResourceQueryContainerPluginInterface;
+
+class EventResourceQueryContainerPluginIterator implements Iterator
+{
+    /**
+     * @var \Spryker\Zed\EventBehavior\Dependency\Plugin\EventResourceQueryContainerPluginInterface
+     */
+    private $plugin;
```
Comment (gechetspr) [inline src/Spryker/Zed/EventBehavior/Business/Model/EventResourceQueryContainerPluginIterator.php]: please change private to protected in all places

File: src/Spryker/Zed/EventBehavior/Business/Model/EventResourceQueryContainerPluginIterator.php [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+    /**
+     * @var array
+     */
+    protected $current = [];
+
+    /**
+     * @var int
+     */
+    protected $chunkSize;
```
Comment (ehsanmx) [inline src/Spryker/Zed/EventBehavior/Business/Model/EventResourceQueryContainerPluginIterator.php]: you can provide a default chunk size here e.g `1000`

File: src/Spryker/Zed/EventBehavior/Business/Model/EventResourceQueryContainerPluginIterator.php [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+    {
+        return $this->index;
+    }
+
+    /**
+     * @return bool
+     */
+    public function valid(): bool
+    {
+        if ($this->plugin->queryData()->exists()) {
```
Comment (ehsanmx) [inline src/Spryker/Zed/EventBehavior/Business/Model/EventResourceQueryContainerPluginIterator.php]: This is wrong:
`$this->plugin->queryData()->exists()` not equal `$this->plugin->queryData()->count() === 0)`

Comment (ehsanmx) [reply src/Spryker/Zed/EventBehavior/Business/Model/EventResourceQueryContainerPluginIterator.php]: you must use `!` for `exists()`, but do you think it's faster by comparing to `count()`?

File: src/Spryker/Zed/EventBehavior/Business/Model/EventResourceQueryContainerPluginIterator.php [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+ * Use of this software requires acceptance of the Evaluation License Agreement. See LICENSE file.
+ */
+
+namespace Spryker\Zed\EventBehavior\Business\Model;
+
+use Iterator;
+use Propel\Runtime\ActiveQuery\ModelCriteria;
+use Spryker\Zed\EventBehavior\Dependency\Plugin\EventResourceQueryContainerPluginInterface;
+
+class EventResourceQueryContainerPluginIterator implements Iterator
```
Comment (ehsanmx) [inline src/Spryker/Zed/EventBehavior/Business/Model/EventResourceQueryContainerPluginIterator.php]: Please apply same fixes for `sync:data`in `spryker/spryker` repository

File: src/Spryker/Zed/EventBehavior/Business/Model/EventResourceQueryContainerPluginIterator.php (hunk 2 above)
Comment (m7moud) [reply src/Spryker/Zed/EventBehavior/Business/Model/EventResourceQueryContainerPluginIterator.php]: It's a required parameter in the constructor.

File: src/Spryker/Zed/EventBehavior/Business/Model/EventResourceQueryContainerPluginIterator.php (hunk 4 above)
Comment (m7moud) [reply src/Spryker/Zed/EventBehavior/Business/Model/EventResourceQueryContainerPluginIterator.php]: Done already :)

**Comment by ehsanmx** [comment]
Two more comments:
- This is a `minor` change
- I'm not sure about legal issues to have external commits here, maybe we can change the author in commits.

File: src/Spryker/Zed/EventBehavior/Business/Model/EventResourceQueryContainerPluginIterator.php (hunk 3 above)
Comment (m7moud) [reply src/Spryker/Zed/EventBehavior/Business/Model/EventResourceQueryContainerPluginIterator.php]: `COUNT` is an aggregation function, `exists` runs the query with `LIIMIT 1` and checks if there's any result for the query, so it's most the time more efficient.

File: src/Spryker/Zed/EventBehavior/Dependency/Facade/EventBehaviorToEventInterface.php [hunk 5]
```diff
@@ -7,17 +7,15 @@
...
 interface EventBehaviorToEventInterface
 {
     /**
      * @param string $eventName
-     * @param \Spryker\Shared\Kernel\Transfer\TransferInterface $eventTransfer
+     * @param \Spryker\Shared\Kernel\Transfer\TransferInterface[] $eventTransfers
      *
      * @return void
      */
-    public function trigger($eventName, TransferInterface $eventTransfer);
```
[the remark this answers is not in the document]
Comment (m7moud) [reply src/Spryker/Zed/EventBehavior/Dependency/Facade/EventBehaviorToEventInterface.php]: I have checked the usage already, but I think I missed this!

File: src/Spryker/Zed/EventBehavior/Business/Model/EventResourceQueryContainerPluginIterator.php (hunk 2 above)
Comment (m7moud) [reply src/Spryker/Zed/EventBehavior/Business/Model/EventResourceQueryContainerPluginIterator.php]: Added to the environment configs.

File: src/Spryker/Zed/EventBehavior/Business/Model/EventResourceRepositoryManager.php [hunk 6]
```diff
@@ -50,7 +49,7 @@ public function __construct(
     }
 
     /**
-     * @param array $plugins
+     * @param \Spryker\Zed\EventBehavior\Dependency\Plugin\EventResourcePluginInterface[] $plugins
      * @param int[] $ids
      *
      * @return void
```
[the remark this answers is not in the document]
Comment (m7moud) [reply src/Spryker/Zed/EventBehavior/Business/Model/EventResourceRepositoryManager.php]: Implemented an iterator instead.

File: src/Spryker/Zed/EventBehavior/Business/Model/EventResourceQueryContainerPluginIterator.php (hunk 3 above)
Comment (m7moud) [reply src/Spryker/Zed/EventBehavior/Business/Model/EventResourceQueryContainerPluginIterator.php]: Removed the extra query, no need for `count` or `exists`.

File: src/Spryker/Zed/EventBehavior/Business/Model/EventResourceRepositoryManager.php [hunk 7]
```diff
@@ -97,58 +84,61 @@ protected function triggerEvents(EventResourceRepositoryPluginInterface $plugin,
...
+        return new EventResourceRepositoryPluginIterator($plugin, $this->chunkSize);
+    }
+
     /**
      * @param \Spryker\Zed\EventBehavior\Dependency\Plugin\EventResourceBulkRepositoryPluginInterface $plugin
      *
      * @return void
      */
-    protected function triggerEventsAllBulk(EventResourceBulkRepositoryPluginInterface $plugin): void
+    protected function processEventsForRepositoryPluginsAllBulk(EventResourceBulkRepositoryPluginInterface $plugin): void
```
Comment (ehsanmx) [inline src/Spryker/Zed/EventBehavior/Business/Model/EventResourceRepositoryManager.php]: Drop `AllBulk`

File: src/Spryker/Zed/EventBehavior/Business/Model/AbstractEventResourcePluginIterator.php [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+/**
+ * Copyright © 2016-present Spryker Systems GmbH. All rights reserved.
+ * Use of this software requires acceptance of the Evaluation License Agreement. See LICENSE file.
+ */
+
+namespace Spryker\Zed\EventBehavior\Business\Model;
+
+use Spryker\Zed\EventBehavior\Dependency\Plugin\EventResourcePluginInterface;
+
+abstract class AbstractEventResourcePluginIterator
```
Comment (ehsanmx) [inline src/Spryker/Zed/EventBehavior/Business/Model/AbstractEventResourcePluginIterator.php]: abstract class must implement `iterator` not concrete classes

**Comment by ehsanmx** [comment]
it’s a minor

File: src/Spryker/Zed/EventBehavior/Business/Model/EventResourceQueryContainerManager.php [hunk 9]
```diff
@@ -70,57 +60,64 @@ public function triggerResourceEvents(array $plugins, array $ids = []): void
...
 
     /**
-     * @param \Spryker\Zed\EventBehavior\Dependency\Plugin\EventResourcePluginInterface $plugin
+     * @param \Spryker\Zed\EventBehavior\Dependency\Plugin\EventResourceQueryContainerPluginInterface $plugin
+     *
+     * @return \Iterator
+     */
+    protected function createEventResourceQueryContainerPluginIterator(EventResourceQueryContainerPluginInterface $plugin): Iterator
+    {
+        return new EventResourceQueryContainerPluginIterator($plugin, $this->chunkSize);
```
Comment (ehsanmx) [inline src/Spryker/Zed/EventBehavior/Business/Model/EventResourceQueryContainerManager.php]: This must be cleaned up with an internal factory which is injected from the constructor, @m7moud please create a ticket for this.

**Pull request merged**